olive/workflows/run/config.py

```diff
@@ -309,6 +291,48 @@ def validate_pass_search(cls, v, values):
         return v
 
 
+def _insert_azureml_client(config, azureml_client):
+    """Insert azureml_client into config if needed.
+
+    Valid cases:
+        1. AzureML resource path config without azureml_client
+        2. AzureML system config without azureml_client_config
+    config is modified in place.
+    """
+    if not isinstance(config, (dict, list)):
+        return
+
+    insert_key, config_type = _needs_aml_client(config)
+
+    if insert_key and not azureml_client:
+        raise ValueError(f"azureml_client is required for {config_type} but not provided.")
+    elif insert_key:
+        set_nested_dict_value(config, insert_key, azureml_client)
+        return
+
+    for value in config.values() if isinstance(config, dict) else config:
+        _insert_azureml_client(value, azureml_client)
+
+
+def _needs_aml_client(config):
+    """Check if azureml_client is needed for the given config.
+
+    Return the path to insert azureml_client and the type of the config.
+    """
+    if not isinstance(config, dict):
+        return None, None
+
+    config_type = config.get("type")
+
+    if config_type in AZUREML_RESOURCE_TYPES and not config.get("config", {}).get("azureml_client"):
+        return ("config", "azureml_client"), "AzureML Resource Path"
```

Comment:
- Does this only cover one level down insertion? Like `config -> azureml_client`.
- Will this be conflicting with the future change if we want to flatten the `config` (for `config.get("config", {})`)?

Reply:
Here we know that we are dealing with azureml resource path (same for aml system in the next condition) so we only look at `config -> azureml_client` because that's where it is expected to be in. Key point note here is that we are only checking to see if this is an azureml related config. the recursion logic is handled by the caller.

Good point about the nesting change (I am starting work on it today). In that case, we can just check directly `config.get("azureml_client")`.

Reply:
Ah yeah I missed the recursion part in `_insert_azureml_client`. Looks good to me.